### PR TITLE
doc: add listen api

### PR DESCRIPTION
**Pull request by artemreyt**
Added api for `listen` to `doc/api.md`.

### CODE REVIEW

File: doc/api.md [hunk 1]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
 
 - `use_body_split`: Boolean, whether to split HTTP(S) request body which does not fit into `shuttle_size` into chunks (which means several trips between HTTP(S) thread and TX thread) or fail request. Note that you are limited by `max_body_len` in any case. Defaults to `False`.
 
+- `listen`: Table or integer or string, depending on a case. We have such term as listener-conf that is an entity describing one listener (IP:port pair).
```
Comment (void234) [inline doc/api.md:82]: 1) Fields are listed in alphabetical order.
2) Description is too large to fit into this list (unreadable), please move it into separate sections like it is done for handlers.
3) Please use Grammarly (free version is enough) or whatever tool use prefer to fix grammar errors.

File: doc/api.md [hunk 2]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+If `listen` is an array (consists only of elements with keys `[1, #listen]`), then
+it is considered as table of listen-confs, in other ways it is considered as one listen-conf (one listener). `listen = 3300`, by default. There're 3 types of listen-conf:
+  + Table. It is the most descriptive way to define listener-conf. Table contains fields:
+    - `addr`: String. IPv4/IPv6 address or hostname. If is empty then 2 listeners with '0.0.0.0' and '::' addresses are created.
```
Comment (void234) [inline doc/api.md:86]: You haven't implemented hostname support (yet?), I just checked. You should fix either code or documentation.

Personally, I do not think there should be support for hostnames here. What do you plan to do if hostname has been changed when the server is already launched? How often would you check?..

File: doc/api.md [hunk 3]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+    - `port`: Integer. Must be in interval [0, 65535]. If is empty, then default 3300 port is used.
+    - `tls`: Array of tables. Each table contains fields `certificate_file` and `certificate_key_file` which are strings with paths of certificate file and private key for a certificate accordingly. If is empty, http protocol is used.
```
Comment (void234) [inline doc/api.md:88]: Both are "files".
You should also emphasize "insecure HTTP".

File: doc/api.md [hunk 4]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+    - `uses_sni`: Boolean. Describes to enable TLS SNI or not. `false`, by default.
```
Comment (void234) [inline doc/api.md:89]: "Specifies whether to enable..."

File: doc/api.md [hunk 5]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+
+    Examples:
+      - ```Lua
+        -- https://foo.com:8080 and https://bar.com:8080
+        -- https://0.0.0.0:8080 won't be accepted due to SNI.
+        {
+          addr = '0.0.0.0',
+          port = 8080,
+          tls = {
+            { certificate_file = 'foo.com_cert.pem', certificate_key_file = 'foo.com_key.pem' },
```
Comment (void234) [inline doc/api.md:99]: I suggest using absolute pathnames in examples. You can use calculated ones if you like - but not relative. Too fragile and insecure.

File: doc/api.md [hunk 6]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+        {
+          addr = '0.0.0.0',
+          port = 8080,
+          tls = { 
+            { certificate_file = 'foo.com_cert.pem', certificate_key_file = 'foo.com_key.pem' }
+          },
+        }
+      }
+      ```
+  There's also an opportunity to listen on the same port but on different IP-addresses and on the same IP-address but on different ports.
```
Comment (void234) [inline doc/api.md:143]: Do you have some special syntax for that? If not, why mention this at all and confuse users?

File: doc/api.md (hunk 1 above)
Comment (artemreyt) [reply doc/api.md]: > 1. Fields are listed in alphabetical order.

Done.

> 2. Description is too large to fit into this list (unreadable), please move it into separate sections like it is done for handlers.

Moved in a separate section.

> 3. Please use Grammarly (free version is enough) or whatever tool use prefer to fix grammar errors.

Thanks, used.

File: doc/api.md (hunk 2 above)
Comment (artemreyt) [reply doc/api.md]: Removed from doc.

File: doc/api.md (hunk 3 above)
Comment (artemreyt) [reply doc/api.md]: Done.

File: doc/api.md (hunk 4 above)
Comment (artemreyt) [reply doc/api.md]: Done.

File: doc/api.md (hunk 5 above)
Comment (artemreyt) [reply doc/api.md]: Done.

File: doc/api.md [hunk 7]
```diff
@@ -79,6 +79,69 @@ httpng module exports the following functions:
...
+        }
+        ```
+      - ```Lua
+        -- Any requests to [::] (https://[::]:3300)
+        {
+          addr = '::',
+          tls = { { certificate_file = 'foo.com_cert.pem', certificate_key_file = 'foo.com_key.pem' } },
+        }
+    
+  + Integer. Defines port to listen by http.
```
[the remark this answers is not in the document]
Comment (artemreyt) [reply doc/api.md]: > "..by HTTP(S) server"

No, if listener-conf is an integer only http protocol can be used.
Changed to "Defines port. In this case will be created listeners with mentioned port on default IP addresses to listen by insecure HTTP protocol."

File: doc/api.md (hunk 6 above)
Comment (artemreyt) [reply doc/api.md]: No, removed from doc.

File: doc/api.md [hunk 8]
```diff
@@ -110,3 +110,71 @@ This is what HTTPNG is about - handling HTTP(S) requests. Handlers are Lua funct
...
+          port = 8080,
+          tls = {
+            { certificate_file = '/path/to/certificate/foo.com_cert.pem', certificate_key_file = '/path/to/key/foo.com_key.pem' },
+            { certificate_file = '/path/to/certificate/bar.com_cert.pem', certificate_key_file = '/path/to/key/bar.com_key.pem' },
+          },
+          uses_sni = true
+        }
+        ```
+      - ```Lua
+        -- Any requests to [::] (https://[::]:3300)
```
Comment (void234) [inline doc/api.md:144]: Not clear enough. You should add a comment that a user's browser would refuse to accept such a page unless `foo.com` in DNS or /etc/hosts points to the HTTPS server's IP.

File: doc/api.md [hunk 9]
```diff
@@ -110,3 +110,71 @@ This is what HTTPNG is about - handling HTTP(S) requests. Handlers are Lua funct
...
+  + Integer. Defines port. In this case, will be created listeners with mentioned port on default IP addresses to listen by an insecure HTTP protocol.
+  + String. Describes URI. Note: HTTPS protocol can't be enabled in this way because `tls` can be defined only if listener-conf is a table.
+
+    Examples:
+      - ```Lua 
+        listen = 'http://0.0.0.0:8080'
+        ```
+      - ```Lua
+        -- listen on foo.com port 3300
+        listen = 'http://foo.com'
```
Comment (void234) [inline doc/api.md:159]: It listens on port 80 (as was requested by @Mons), not on 3300.

File: doc/api.md (hunk 1 above)
Comment (void234) [reply doc/api.md]: > > 1. Description is too large to fit into this list (unreadable), please move it into separate sections like it is done for handlers.
> 
> Moved in a separate section.

You still should note the existence of this field with "...for more details please see a section below".
 
> > 1. Please use Grammarly (free version is enough) or whatever tool use prefer to fix grammar errors.
> 
> Thanks, used.

BTW, you can use https://markdown-it.github.io/ with Grammarly if you like.

**Review by void234 — CHANGES_REQUESTED** [review]
Please see comments above.

File: doc/api.md (hunk 8 above)
Comment (artemreyt) [reply doc/api.md:144]: > a user's browser would refuse to accept such a page unless `foo.com` in DNS or /etc/hosts points to the HTTPS server's IP.

I think it's obvious. It's basic of client-server architecture, isn't it?

Comment (void234) [reply doc/api.md:144]: > > a user's browser would refuse to accept such a page unless `foo.com` in DNS or /etc/hosts points to the HTTPS server's IP.
> 
> I think it's obvious. It's the basics of client-server architecture, isn't it? The client should know IP-address of the server. The domain mentioned in the certificate doesn't matter in this case.

It does. If DNS says "foo.tarantool.io 188.93.56.54" but the certificate says "bar.tarantool.io" - a user browser would refuse to show the web page, that's the idea.

File: doc/api.md (hunk 9 above)
Comment (artemreyt) [reply doc/api.md:160]: Changed to 80 port.

File: doc/api.md (hunk 8 above)
Comment (artemreyt) [reply doc/api.md]: > a user's browser would refuse to accept such a page unless foo.com in DNS or /etc/hosts points to the HTTPS server's IP.

Ok. Added this note.

File: doc/api.md [hunk 10]
```diff
@@ -110,3 +110,72 @@ This is what HTTPNG is about - handling HTTP(S) requests. Handlers are Lua funct
...
+
+`httpng.cfg{ listen = <...> }`
+
+The way to specify server listeners (IP:port pairs) in HTTPNG is to pass a value with a key called `listen` to the table which is the first argument to `httpng.cfg()` function. There are some rules on how to set `listen`.
+
+`listen`: Table or integer or string, depending on a case. We have such a term as "listener-conf" which is an entity describing one listener (IP:port pair).
+If `listen` is an array (consists only of elements with keys `[1, #listen]`), then
+it is considered as a table of listener-confs, in other ways it is considered as one listener-conf (one listener). `listen = 3300`, by default. There're 3 types of listener-conf:
+  + Table. It is the most descriptive way to define listener-conf. The table contains fields:
+    - `addr`: String. IPv4/IPv6 address. If it is empty then 2 listeners with '0.0.0.0' and '::' addresses are created.
```
Comment (void234) [inline doc/api.md:124]: `addr = ''` causes "Can't parse IP address" so this description is not accurate.

File: doc/api.md [hunk 11]
```diff
@@ -110,3 +110,72 @@ This is what HTTPNG is about - handling HTTP(S) requests. Handlers are Lua funct
...
+          tls = {
+            { certificate_file = '/path/to/certificate/foo.com_cert.pem', certificate_key_file = '/path/to/key/foo.com_key.pem' },
+            { certificate_file = '/path/to/certificate/bar.com_cert.pem', certificate_key_file = '/path/to/key/bar.com_key.pem' },
+          },
+          uses_sni = true
+        }
+        ```
+      - ```Lua
+        -- Any requests to [::] (https://[::]:3300).
+        -- Note: a user's browser would refuse to accept such a page unless foo.com in DNS or /etc/hosts points to the HTTPS server's IP.
```
Comment (void234) [inline doc/api.md:145]: "Note: a user's browser would refuse to accept such a page unless it attempts to access the website from this certificate (`foo.com` in this example)".

File: doc/api.md [hunk 12]
```diff
@@ -110,3 +110,72 @@ This is what HTTPNG is about - handling HTTP(S) requests. Handlers are Lua funct
...
+        {
+          addr = '::',
+          tls = { { certificate_file = 'path/to/certificate/foo.com_cert.pem', certificate_key_file = '/path/to/key/foo.com_key.pem' } },
+        }
+    
+  + Integer. Defines port. In this case, will be created listeners with mentioned port on default IP addresses to listen by an insecure HTTP protocol.
+  + String. Describes URI. Note: HTTPS protocol can't be enabled in this way because `tls` can be defined only if listener-conf is a table.
```
Comment (void234) [inline doc/api.md:152]: `listen = { 'https://foo.tarantool.io' }` enables insecure HTTP on port 443. This is quite a bad idea. I believe you should fix the code (throw error) in a separate PR.

**Review by void234 — CHANGES_REQUESTED** [review]
Please see the comments above. We are nearly there.

File: doc/api.md (hunk 10 above)
Comment (artemreyt) [reply doc/api.md]: Changed `empty` to `nil`.

File: doc/api.md (hunk 11 above)
Comment (artemreyt) [reply doc/api.md]: Done.

File: doc/api.md (hunk 12 above)
Comment (artemreyt) [reply doc/api.md:154]: Will be done in a separate PR

**Pull request merged**